virttest/utils_misc.py

```diff
@@ -2818,199 +2818,314 @@ def get_uptime(session=None):
     return float(uptime.split()[0])
 
 
-def list_linux_guest_disks(session, partition=False):
+def get_linux_disks(session, partition=False):
     """
-    List all disks OR disks with no partition in linux guest.
+    List all disks or disks with no partition.
 
     :param session: session object to guest
     :param partition: if true, list all disks; otherwise,
                       list only disks with no partition.
-    :return: the disks set.
-    """
-    cmd = "ls /dev/[vhs]d*"
+    :return: the disks dict.
+             e.g. {kname: [kname, size, type, serial, wwn]}
+    """
+    list_disk_cmd = "lsblk -o KNAME,SIZE,TYPE"
+    output = session.cmd_output_safe(list_disk_cmd)
+    devs = output.splitlines()
+    disks_dict = {}
+    part_dict = {}
+    for dev in devs:
+        dev = dev.split()
+        if "disk" in dev:
+            disks_dict[dev[0]] = dev
+        if "part" in dev:
+            part_dict[dev[0]] = dev
     if not partition:
-        cmd = "%s | grep -v [0-9]$" % cmd
-    status, output = session.cmd_status_output(cmd)
-    if status != 0:
-        raise exceptions.TestFail("Get disks failed with output %s" % output)
-    return set(output.split())
+        for disk in disks_dict.keys():
+            if disk in ",".join(part_dict.keys()):
+                disks_dict.pop(disk)
+    else:
+        for part in part_dict.keys():
+            disks_dict[part] = part_dict[part]
+    get_info_cmd = "udevadm info -q property -p /sys/block/%s"
+    for kname in disks_dict.keys():
+        output = session.cmd_output_safe(get_info_cmd % kname)
+        serial = re.findall(r"(?<=ID_SERIAL=)(.*)", output, re.M)
+        if serial:
+            disks_dict.get(kname).extend(serial)
```

Comment:
I think if `serial` is not matched, the result list should append `None`.
Same for `wwn`

Reply:
No need to append None. In function \"format_linux_disk\", 
all_disks_did = get_linux_disks(session, partition).values(), it is a list.
e.g. [kname, size, type, serial, wwn], [kname, size, type, wwn], [kname, size, type, serial] or [kname, size, type]
In for loop, it will be judged by code \"not in\".  
Details as below:
for disk_did in all_disks_did:
        if did not in disk_did:

Reply:
did maybe kname, serial, or wwn.  disk_did maybe [kname, size, type, serial, wwn], [kname, size, type, wwn], [kname, size, type, serial] or [kname, size, type]

Reply:
My concern is if someone use it directly, the return value is not easy to decide which is which as the items in the list are varies.